core/che-core-api-core/src/main/java/org/eclipse/che/api/core/util/Version.java

```diff
@@ -0,0 +1,246 @@
+import java.util.regex.Pattern;
+
+import static java.lang.Integer.parseInt;
+import static java.lang.String.format;
+import static java.util.regex.Pattern.compile;
+
+/**
+ * Version format:
+ * {MajorVersion}.{MinorVersion}.{BugFixVersion}.{HotFixVersion}-M{MilestoneVersion}-beta-{BetaVersion}-RC-SNAPSHOT.
+ * </p>
+ * Mandatory parts are only major, minor and bug fix versions.
+ *
+ * @see #toString()
+ * @see #VERSION
+ *
+ * @author Anatoliy Bazko
+ */
+public class Version implements Comparable<Version> {
+
+    private static final String MILESTONE_VERSION_SUFFIX = "-M";
+    private static final String BETA_VERSION_SUFFIX      = "-beta-";
+
+    private static final Pattern VERSION =
+            compile("^(?<major>0|[1-9]+[0-9]*)\\.(?<minor>0|[1-9]+[0-9]*)\\.(?<bugfix>0|[1-9]+[0-9]*)(?<hotfix>\\.0|\\.[1-9]+[0-9]*)?" +
+                    "(?<milestone>" + MILESTONE_VERSION_SUFFIX + "[1-9]+[0-9]*)?" +
+                    "(?<beta>" + BETA_VERSION_SUFFIX + "[1-9]+[0-9]*)?" +
+                    "(?<rc>-RC)?" +
+                    "(?<snapshot>-SNAPSHOT)?$");
+
+    private final int     major;
+    private final int     minor;
+    private final int     bugFix;
+    private final int     hotFix;
+    private final int     milestone;
+    private final int     beta;
+    private final boolean snapshot;
+    private final boolean rc;
+
+    public Version(int major,
+                   int minor,
+                   int bugFix,
+                   int hotFix,
+                   int milestone,
+                   int beta,
+                   boolean rc,
+                   boolean snapshot) {
+        this.major = major;
+        this.minor = minor;
+        this.bugFix = bugFix;
+        this.hotFix = hotFix;
+        this.milestone = milestone;
+        this.beta = beta;
+        this.snapshot = snapshot;
+        this.rc = rc;
+    }
+
+    /**
+     * Checks if version format is valid.
+     *
+     * @throws IllegalArgumentException
+     *      if {@code version} format is wrong
+     */
+    public static void validate(@NotNull String version) throws IllegalArgumentException {
+        if (!VERSION.matcher(version).matches()) {
+            throw new IllegalArgumentException(format("Illegal version '%s' format.", version));
+        }
+    }
+
+    /**
+     * Parse version in string representation.
+     *
+     * @throws IllegalArgumentException
+     *      if {@code version} format is wrong
+     */
+    public static Version parse(@NotNull String version) throws IllegalArgumentException {
+        Matcher matcher = VERSION.matcher(version);
+        if (!matcher.find()) {
+            throw new IllegalArgumentException(format("Illegal version '%s' format.", version));
+        }
+
+        int hotFix = 0;
+        int milestone = 0;
+        int beta = 0;
+
+        String hotFixGroup = matcher.group("hotfix");
+        if (hotFixGroup != null) {
+            hotFix = parseInt(hotFixGroup.substring(1));
```

Comment:
You can rewrite Pattern to avoid evaluating substring while parsing hotFix, milestone and beta versions.